executor/adapter.go

```diff
@@ -595,6 +595,19 @@ func (a *ExecStmt) handlePessimisticLockError(ctx context.Context, err error) (E
 		if conflictCommitTS > forUpdateTS {
 			newForUpdateTS = conflictCommitTS
 		}
+	} else if terror.ErrorEqual(err, tikv.ErrLockFailNoWait) {
+		// for nowait, when ErrLock happened, ErrLockFailNoWait will be returned, and in the same txn
+		// the select for updateTs must be updated, otherwise there maybe rollback problem.
+		// begin;  select for update key1(here ErrLocked or other errors(or max_execution_time like util),
+		//         key1 lock not get and async rollback key1 is raised)
+		//         select for update key1 again(this time lock succ(maybe lock released by others))
+		//         the async rollback operation rollbacked the lock just acquired
+		newForUpdateTS, tsErr := a.Ctx.GetStore().GetOracle().GetTimestamp(ctx)
+		if tsErr != nil {
+			return nil, tsErr
+		}
+		txnCtx.SetForUpdateTS(newForUpdateTS)
+		return nil, err
```

Comment:
return err?

Reply:
Here do update `updateTS` in the same txn, and report `ErrLocked` to the user when `no wait` used,  to avoid the async rollback problem described below

Reply:
Get it!